Unix/http/httpauth.c

```diff
@@ -0,0 +1,1441 @@
+
+#if AUTHORIZATION
+    static const char *RESPONSE_HEADER_BAD_REQUEST =
+        "HTTP/1.1 400 Bad Request\r\n" "Content-Length: 0\r\n" "\r\n";
+    OM_uint32 flags = 0;
+    const char *protocol_p = NULL;
+#endif
+
+    unsigned char *auth_response = NULL;
+    int response_len = 0;
+
+    if (IsAuthCallsIgnored()) {
+        handler->httpErrorCode = 0; // We let the transaction set the error code
+        handler->isAuthorised = TRUE;
+        return TRUE;
+    }
+
+    if (!headers) {
+        handler->httpErrorCode = HTTP_ERROR_CODE_INTERNAL_SERVER_ERROR;
+        goto Done;
+    }
+
+    if (Strncasecmp(headers->authorization, AUTHENTICATION_BASIC, AUTHENTICATION_BASIC_LENGTH) == 0) {
+        if (!headers->username || !headers->password
+            || 0 != AuthenticateUser(headers->username, headers->password)) {
+            handler->httpErrorCode = HTTP_ERROR_CODE_UNAUTHORIZED;
+            auth_response = (unsigned char *)RESPONSE_HEADER_UNAUTH_FMT;
+            response_len = strlen(RESPONSE_HEADER_UNAUTH_FMT);
+
+            // trace msg here
+            handler->authFailed = TRUE;
+
+            _SendAuthResponse(handler, auth_response, response_len);
+            return FALSE;
+        }
+
+        if (0 != LookupUser(headers->username, &handler->authInfo.uid, &handler->authInfo.gid)) {
+            trace_GetUserUidGid_Failed(headers->username);
+
+            handler->httpErrorCode = HTTP_ERROR_CODE_UNAUTHORIZED;
+            auth_response = (unsigned char *)RESPONSE_HEADER_UNAUTH_FMT;
+            response_len = strlen(RESPONSE_HEADER_UNAUTH_FMT);
+
+            handler->authFailed = TRUE;
+            _SendAuthResponse(handler, auth_response, response_len);
+            return FALSE;
+        }
+
+        handler->httpErrorCode = 0; // Let the request do the error code
+        handler->isAuthorised = TRUE;
+        return TRUE;
+    }
+#ifdef AUTHORIZATION
+    else {
+        const gss_OID_desc mech_krb5 = { 9, "\052\206\110\206\367\022\001\002\002" };
+        const gss_OID_desc mech_spnego = { 6, "\053\006\001\005\005\002" };
+        const gss_OID_desc mech_iakerb = { 6, "\053\006\001\005\002\005" };
+        //const gss_OID_desc mech_ntlm   = {10, "\x2b\x06\x01\x04\x01\x82\x37\x02\x02\x0a" };
+        // gss_OID_set_desc mechset_krb5 = { 1, &mech_krb5 };
+        // gss_OID_set_desc mechset_iakerb = { 1, &mech_iakerb };
+        const gss_OID_set_desc mechset_spnego = { 1, (gss_OID) & mech_spnego };
+
+        const gss_OID mechset_krb5_elems[] = { (gss_OID const)&mech_krb5,
+            (gss_OID const)&mech_iakerb
+        };
+
+        const gss_OID_set_desc mechset_krb5 = { 2, (gss_OID) mechset_krb5_elems };
+
+        OM_uint32 maj_stat, min_stat;
+        gss_ctx_id_t context_hdl = GSS_C_NO_CONTEXT;
+        gss_buffer_desc input_token, output_token;
+        gss_OID_set mechset = NULL;
+
+        // Ensure the GSS lib is loaded
+
+        switch (_g_gssState.gssLibLoaded)
```

Comment:
really need to use Once_Invoke otherwise there will be race conditions in this code